syntax = "proto3";

package rpc;

service ExperimentManagerRPC {
    rpc CreateTable(TableName) returns (TableKey);
    rpc SendText(Text) returns (SendReply);
    rpc SendScalar(Scalar) returns (SendReply);
    rpc SendImage(stream Binary) returns (SendReply);
    rpc SendBinaryTensor(stream Binary) returns (SendReply);
    rpc SendObj(stream Binary) returns (SendReply);
}

message TableName {
    string primary = 1;
    string secondary = 2;
    string src = 4;
    double time = 5;
}

message TableKey {
    int32 key = 1;
    double time = 2;
}

message Text {
    int32 key = 1;
    string tag = 2;
    string src = 3;
    string text = 4;
    int32 step = 5;
    double time = 6;
}

message Scalar {
    int32 key = 1;
    string tag = 2;
    string src = 3;
    oneof ScalarType {
        int64 op_int = 4;
        float op_float = 6;
    }
    int32 step = 7;
    double time = 8;
}

message Binary {
    int32 key = 1;
    string tag = 2;
    string src = 3;
    bool tensor = 4;
    bytes blocks = 5;
    int32 step = 6;
    double time = 7;
}

message SendReply {
    int32 status = 1;
    double time = 2;
}